### ISSUE TITLE
performance: GetProperties() only once during runtime

### ISSUE THREAD

**Issue by penCsharpener** [sqlkata/querybuilder#245]
When running inserts the code checks for the property attributes Ignore and Column. This is done every single time an insert statement is run. Dapper.FluentMapper does this more efficiently by getting mappings once as early as possible during runtime which reduces overhead. This requires some form of static dictionary of mappings and types.

**Comment by asherber** [comment]
I'm all in favor of flexibility, but I would caution against trying to get this library to do everything. IMO, the strength of the library is in building queries; the fact that it can do some execution of those queries through Dapper is a nice bonus.

When we start getting into these kinds of attributes on properties to help with mapping, the library starts getting into micro-ORM territory -- and there are already several of those that have dealt with these issues. (

Personally, I use PetaPoco; it includes a rudimentary SQL builder, but I wrote a small library that lets it use SqlKata instead. PetaPoco already knows how to map to/from objects (and to cache object properties), so I can mix and match with SqlKata as appropriate, building complex select queries with SqlKata but using PetaPoco's built-in insert-from-object capabilities.)

So I would caution against straying too far from the core functionality of this library.

**Comment by for7raid** [comment]
Fully agree with @asherber

**Comment by penCsharpener** [comment]
These additions are concerning the SqlKata.Extensions which is a separate nuget package as well just as Dapper has various extensions. I don't see a problem with building more into the .Extensions part. @asherber I'm assuming you are using vanilla SqlKata without the Extensions with your PetaPoco library, right? So then what is done in the Extensions is not affecting the core idea of SqlKata.

**Comment by asherber** [comment]
@penCsharpener I'm a little confused. The `Ignore` and `Column` attributes are part of the core SqlKata library, and since the discussion is taking place here, I assumed it was about more functionality to be added here. Where is the repository for SqlKata.Extensions?

**Comment by ahmad-moussawi** [comment]
Actually, they live inside the same Github repo, but they are two different projects (two NuGet packages), but in general, I agree that we should consider additional features carefully to not fall in the ORM trap.

In this issue `Column` and `Key` attributes are really needed since SqlKata supports inserting/updating object.

**Comment by penCsharpener** [comment]
Yes, my mistake. Extensions covers everything around executing the queries built. The attributes are part of SqlKata without Extensions. But as @ahmad-moussawi said, they still have their place.

**Comment by rwood** [comment]
I noticed that there was a cache dictionary already while making some changes I needed elsewhere.  Is this a sufficient change to handle the cache?  Do we need to worry about pruning the cache over time, or limiting its size?
https://github.com/sqlkata/querybuilder/pull/259
